### DISCUSSION TITLE
postgres defaultValue isnt applied

### DISCUSSION THREAD

**Discussion by ryanleecode** [payloadcms/payload#6691] [Q&A]
### Payload Version

3.0.0-beta.43

### Node Version

20.12.2

### Next.js Version

15.0.0-rc.0

### Describe the Bug

When you change a `text` field to `required` and set the `defaultValue` the defaultValue isn't applied So you will just get data loss warnings.

`DATA LOSS WARNING: Possible data loss detected if schema is pushed.`

### Reproduction Steps

.

**Comment by DanRibbens** [comment]
This is a situation where you need to create a migration. Payload handles `defaultValues` entirely in the application layer and not through to the database. Since `defaultValue` can be a function that return values it has to be run at each create. Another challenge is that table structures aren't always simple columns being added like a Payload field that needs a join table such as arrays or fields with `hasMany: true`, for these we can't take a defaultValue like `[1, 2, 3]` and apply it as three inserts into a numbers table as an example.

Instead of trying to handle all the situations and only being able to cover some of them automatically, we let that fall into the user migration responsibility for any cases.

All you need is to do:
1. `payload create:migration add_required_foo_table_bars` or a different name.
2. In the `up` function you should already have the schema change written for you.
3. Above the schema change, add a function to store values for all existing rows:
```ts
import { sql } from 'drizzle-orm'

// in your up function:
await payload.db.drizzle.execute(sql`UPDATE bars SET foo = 'newDefault' WHERE foo IS NULL;`)
```
4. Test your migration locally before deploying and running on production.

**Comment by ryanleecode** [comment]
Trade off is worth it for simple fields IMO. The dev experience just sucks whenever you make a minor change right now. A custom migration script can be made for more complex stuff

**Comment by ryanleecode** [comment]
This has proven highly useful to me

```diff
--- a/packages/db-postgres/src/schema/traverseFields.ts
+++ b/packages/db-postgres/src/schema/traverseFields.ts
@@ -168,14 +168,14 @@ export const traverseFields = ({
             )
           }
         } else {
-          targetTable[fieldName] = varchar(columnName)
+          targetTable[fieldName] = varchar(columnName).default(field.defaultValue)
         }
         break
       }
       case 'email':
       case 'code':
       case 'textarea': {
-        targetTable[fieldName] = varchar(columnName)
+        targetTable[fieldName] = varchar(columnName).default(field.defaultValue)
         break
       }
 
@@ -197,14 +197,14 @@ export const traverseFields = ({
             )
           }
         } else {
-          targetTable[fieldName] = numeric(columnName)
+          targetTable[fieldName] = numeric(columnName).default(field.defaultValue)
         }
         break
       }
 
       case 'richText':
       case 'json': {
-        targetTable[fieldName] = jsonb(columnName)
+        targetTable[fieldName] = jsonb(columnName).default(field.defaultValue)
         break
       }
 
@@ -213,7 +213,7 @@ export const traverseFields = ({
           mode: 'string',
           precision: 3,
           withTimezone: true,
-        })
+        }).default(field.defaultValue)
         break
       }
```

**Comment by r1tsuu** [comment]
PR: https://github.com/payloadcms/payload/pull/6983

**Comment by DanRibbens** [comment]
I had to do this as a separate PR. https://github.com/payloadcms/payload/pull/7368
